htmresearch/frameworks/nlp/imbu.py

```diff
@@ -0,0 +1,214 @@
+
+    kwargs.update(**self._defaultModelFactoryKwargs())
+
+    if modelType == "CioWordFingerprint":
+      kwargs.update(retina=kwargs.get("retina"),
+                    apiKey=kwargs.get("apiKey"),
+                    fingerprintType=EncoderTypes.word,
+                    cacheRoot=self.cacheRoot)
+
+    elif modelType == "CioDocumentFingerprint":
+      kwargs.update(retina=kwargs.get("retina"),
+                    apiKey=kwargs.get("apiKey"),
+                    fingerprintType=EncoderTypes.document,
+                    cacheRoot=self.cacheRoot)
+
+    elif modelType == "HTMNetwork":
+      kwargs.update(networkConfig=_loadNetworkConfig(),
+                    inputFilePath=None,
+                    prepData=False,
+                    retinaScaling=1.0)
+
+    model = createModel(modelType, **kwargs)
+
+
+    model.verbosity = 0
+
+    return model
+
+
+  def loadModel(self, *modelFactoryArgs, **modelFactoryKwargs):
+
+    try:
+      model = ClassificationModel.load(self.loadPath)
+    except IOError as exc:
+      model = self.modelFactory(*modelFactoryArgs, **modelFactoryKwargs)
+      self.train(model)
+
+    return model
+
+
+  def _loadTrainingData(self):
+    return readCSV(self.dataPath, numLabels=0)
+
+
+  def train(self, model):
+    for seqId, (text, _, _) in enumerate(self.dataDict.values()):
+      model.trainDocument(text, [seqId], seqId)
+
+    self.save(model)
+
+
+  @staticmethod
+  def query(model, query, returnDetailedResults=True, sortResults=True):
+    return model.inferDocument(query,
+                               returnDetailedResults=returnDetailedResults,
+                               sortResults=sortResults)
+
+
+  def save(self, model):
+    model.save(self.savePath)
+
+
+
+def main():
+  parser = argparse.ArgumentParser()
+
+  parser.add_argument("--cacheRoot",
+                      help="Root directory in which to cache encodings")
+  parser.add_argument("--modelSimilarityMetric",
+                      default=ImbuModels.defaultSimilarityMetric,
+                      help="Classifier metric")
+  parser.add_argument("--dataPath",
+                      help="Path to data CSV; samples must be in column w/ "
+                           "header 'Sample'; see readCSV() for more.",
+                      required=True)
+  parser.add_argument("--modelName",
+                      choices=ClassificationModelTypes.getTypes(),
+                      default=ImbuModels.defaultModelType,
+                      type=str,
+                      help="Name of model class. Also used for model results "
+                           "directory and pickle checkpoint.")
+  parser.add_argument("--loadPath",
+                      default="",
+                      type=str,
+                      help="Path from which to load a serialized model.")
+  parser.add_argument("--savePath",
+                      default="",
+                      type=str,
+                      help="Path to save the serialized model.")
```

Comment:
Please add details explaining what is expected of the `ClassificationModel`: https://github.com/numenta/nupic.research/blob/master/htmresearch/frameworks/nlp/classification_model.py#L253-L259